ding/model/template/ebm.py

```diff
@@ -0,0 +1,488 @@
+"""
+Vanilla DFO and EBM are adapted from https://github.com/kevinzakka/ibc.
+MCMC is adapted from https://github.com/google-research/ibc.
+"""
+from typing import Callable, Tuple
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from abc import ABC, abstractmethod
+
+from ding.utils import MODEL_REGISTRY, STOCHASTIC_OPTIMIZER_REGISTRY
+from ding.torch_utils import unsqueeze_repeat
+from ding.model.wrapper import IModelWrapper
+from ..common import RegressionHead
+
+
+def create_stochastic_optimizer(device, stochastic_optimizer_config):
+    return STOCHASTIC_OPTIMIZER_REGISTRY.build(
+        stochastic_optimizer_config.pop("type"), device=device, **stochastic_optimizer_config
+    )
+
+
+def no_ebm_grad():
+    """Wrapper that disables energy based model gradients"""
+
+    def ebm_disable_grad_wrapper(func: Callable):
+
+        def wrapper(*args, **kwargs):
+            ebm = args[-1]
+            assert isinstance(ebm, (IModelWrapper, nn.Module)),\
+                   'Make sure ebm is the last positional arguments.'
+            ebm.requires_grad_(False)
+            result = func(*args, **kwargs)
+            ebm.requires_grad_(True)
+            return result
+
+        return wrapper
+
+    return ebm_disable_grad_wrapper
+
+
+class StochasticOptimizer(ABC):
+
+    def _sample(self, obs: torch.Tensor, num_samples: int) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Helper method for drawing action samples from the uniform random distribution
+        and tiling observations to the same shape as action samples.
+        obs: (B, O)
+        return: (B, N, O), (B, N, A).
+        """
+        action_bounds = self.action_bounds.cpu().numpy()
+        size = (obs.shape[0], num_samples, action_bounds.shape[1])
+        action_samples = np.random.uniform(action_bounds[0, :], action_bounds[1, :], size=size)
```

Comment:
why not use `torch.rand`?

Reply:
I can not directly set bounds in `torch.rand` and I have to apply affine transformation manually, which is not as decent and readable as numpy method.

Reply:
the operation in `numpy.random.uniform` is also a kind of  affine transformation, i.e., `low + (high-low) * random_sample()`,  I think keep all the torch implementation is more important

Reply:
OK, changed